### PR TITLE
AMM Dex

### LINKED ISSUE #241: [FEATURE] Implement AMM Spec
## dex.go - implements AMM style atomic swaps between root & nested chains
--
## Overview

This file is designed to facilitate a Uniswap V2 style AMM token swap
between a root and a nested chain.

The AMM executes **batched limit orders** and supports **one-sided liquidity deposits and withdrawals**.

Batches are processed **atomically between chains** via receipts and a strict locking protocol, ensuring deterministic outcomes and correct LP point allocation.

## 1. Core Idea: Uniswap AMM formula

At its heart, the Uniswap invariant is:

```
x * y = k
```

- `x` = reserve of token A (add-side)
- `y` = reserve of token B (distribute-side)
- `k` = constant product

In `dex.go`:

```go
dY = (amountInWithFee * y) / (x*1000 + amountInWithFee)
```

This is literally the **Uniswap V2 formula** with fee `0.3%`. Every swap still satisfies `x * y` invariant (ignoring rounding errors).

---

## 2. Virtual Pool Mirrors Counterparty State

- Each chain maintains a `virtualPool` based on the **other chain’s last `PoolSize`**.
- When processing orders, it treats this as the effective reserve (`x`) for the “add” token.
- Trades are applied **against local liquidity (`y`)**, but the invariant is computed against the mirrored counterparty pool.

---

## 3. LP Points Are Standard Pro-Rata

Liquidity addition/withdrawal follows the classic Uniswap V2 integer math:

```go
ΔL = L * (√((x+d)*y) - √(x*y)) / √(x*y)
```

- LP points proportional to **square root of the pool product** (`√k`)
- One-sided deposits work because counterparty’s pool size is included (`x` = remote PoolSize)
- Withdrawals are proportional to **LP share**

---

## 4. Batches Don’t Break Uniswap Logic

- Batches aggregate multiple orders → atomic execution, prevent reentrancy, ensure deterministic receipts.
- Within a batch:
    - Orders processed sequentially (pseudorandomly to prevent front-running)
    - Each trade updates `x` and `y` (virtual and local reserves)
    - `dY` computation is identical to Uniswap

[x] Batching = different implementation, same economic outcome.

---

## 5. Cross-Chain Receipts Ensure Determinism

- Nested reads Root’s locked batch and vice versa.
- Trades only finalize after receipts → both sides see the same LP state.
- Failed orders refunded → invariant preserved.

---

## 6. Fees Are Applied Identically

```go
amountInWithFee = dX * 997
dY = (amountInWithFee * y) / (x*1000 + amountInWithFee)
```

- Standard 0.3% Uniswap V2 fee
- Deducted from incoming token → same effect on `x*y` and LP gains

---

## Terminology & Symbols

- **Tokens**
    - `nested` — token on Nested Chain
    - `root` — token on Root Chain
- **Pools**
    - `HoldingPool[c]` — escrow for pending orders and deposits on chain `c`
    - `LiquidityPool[c]` — settled funds ready for swaps on chain `c`
- **Batches**
    - `NextBatch[c]` — collects new activity
    - `LockedBatch[c]` — awaiting cross-chain receipt processing
- **Orders & Liquidity**
    - `DexLimitOrder { Address, AmountForSale, RequestedAmount }`
        - `RequestedAmount` = minimum acceptable output
    - `DexLiquidityDeposit { Address, Amount }` (one-sided)
    - `DexLiquidityWithdraw { Address, Percent }`
- **LP Accounting**
    - `L[c]` = total LP points on chain `c`
    - `P[c, addr]` = LP points for address `addr`
- **AMM Pricing**
    - Let `x` = add-side pool (snapshot from counterparty)
    - Let `y` = distribute-side pool (local)
    - Invariant: `k = x * y`
- **Fees**
    - `f_taker = 0.003` (default, 30 bps)

---

## State & Storage

For each chain `c`:

- `HoldingPool[c].amount`
- `LiquidityPool[c].amount`
- LP ledger: `L[c]`, `P[c, addr]`
- `NextBatch[c]`:
    - `Orders`, `Deposits`, `Withdraws`
    - `PoolSize` snapshot of `LiquidityPool[c]`
- `LockedBatch[c]`:
    - Last sent batch awaiting receipts
    - Contains per-order `Receipts` once processed by counterparty

---

## Transaction Types

1. **DexLimitOrderTx(chain=c)**
    - Move `AmountForSale` → `HoldingPool[c]`
    - Append order to `NextBatch[c].Orders`
2. **DexLiquidityDeposit(chain=c)**
    - Move `Amount` → `HoldingPool[c]`
    - Append to `NextBatch[c].Deposits`
3. **DexLiquidityWithdraw(chain=c)**
    - Append to `NextBatch[c].Withdraws` (no immediate token move)
4. **certificateResultTx(from=c to=c')**
    - Encodes `LockedBatch[c]` with `PoolSize` and per-order receipts

---

## Batch Processing Cycle

Each chain processes its **locked batch** on a 'trigger'

**Trigger:**

- Nested Chain: `begin_block` after processing previous `certificateResult`
- Root Chain: `deliver_tx` on inbound `certificateResultTx`

Below are the **mirrored steps** that a nested and root chain takes to facilitate the AMM swaps.

**Steps (per chain `c`, counterparty `c'`):**

1. **Process Inbound Receipts (LockedBatch[c])**
    - If receipts incomplete → abort (retry next trigger)
    - Else:
        - Orders succeeded → move tokens from `HoldingPool` → `LiquidityPool`
        - Orders failed → refund from `HoldingPool` → seller
    - Process LockedBatch deposits & withdrawals

2. **Process Inbound Sell Orders (from counterparty batch)**
    - Snapshot: `x = remoteBatch.PoolSize`, `y = LiquidityPool[c].amount`
    - Sort orders pseudorandomly
    - For each order (`dX`):
        - `amountInWithFee = dX * 997`
        - `dY = (amountInWithFee * y) / (x*1000 + amountInWithFee)`
        - If `dY >= RequestedAmount` → accept, update `x += dX`, `y -= dY`, record receipt
        - Else reject

3. **Process Local Liquidity**
    - Withdrawals → burn points, distribute tokens pro-rata
    - Deposits → pro-rata mint LP points based on `(x + deposits) * y`

4. **Rotate Batches**
    - `NextBatch[c] → LockedBatch[c]` (embed receipts)
    - `PoolSize = LiquidityPool[c].amount`
    - Clear `NextBatch[c]`

**Safety Notes:**
- Locked batches prevent reentrancy; no new orders processed until receipts arrive
- Timeouts allow safe retry/failure handling

---

## One-Sided Liquidity Logic

***Two-chain LP accounting**
- Mirror liquidity ledger on both chains for symmetry
- Outbound deposits/withdraws: update ledger + move tokens (use dX from order clearing)
- Inbound deposits/withdraws: update ledger but only token movement for withdraws

**Deposits**
- Let `x = batch.PoolSize`, `y = counterPoolSize`, `d = deposit amount`, `L = total LP points`
- If `L == 0`: initialize with `L = √(x*y)`
- Compute:
    - `oldK = x * y`
    - `newK = (x + d) * y`
    - `ΔL = L * (√newK - √oldK) / √oldK`
- Distribute `ΔL` pro-rata to depositors
- Funds move from Holding → Liquidity

**Withdrawals**
- Burn LP points proportionally
- Distribute tokens according to share of reserves
- Both real (`yPoolSize`) and virtual (`batch.PoolSize`) reserves used

---

```mermaid
sequenceDiagram
    participant UserNested as Nested User
    participant UserRoot as Root User
    participant NestedChain
    participant NestedHolding as Nested HoldingPool
    participant NestedLiquidity as Nested LiquidityPool
    participant NestedNext as Nested NextBatch
    participant NestedLocked as Nested LockedBatch
    participant RootChain
    participant RootHolding as Root HoldingPool
    participant RootLiquidity as Root LiquidityPool
    participant RootNext as Root NextBatch
    participant RootLocked as Root LockedBatch

    %% User submissions
    UserNested->>NestedNext: Submit Limit Order / Liquidity Deposit / Withdraw
    NestedNext->>NestedHolding: Move funds into HoldingPool

    UserRoot->>RootNext: Submit Limit Order / Liquidity Deposit / Withdraw
    RootNext->>RootHolding: Move funds into HoldingPool

    %% Nested reads Root LockedBatch receipts
    NestedChain->>RootLocked: Read receipts to unlock trades
    RootLocked-->>NestedChain: Receipt data

    %% Nested Batch Processing
    NestedChain->>NestedLocked: Process Inbound Receipts
    NestedLocked->>NestedLiquidity: Success → Move funds
    NestedLocked->>UserNested: Fail → Refund

    NestedChain->>NestedLocked: Process Local Liquidity
    NestedLocked->>NestedLiquidity: Deposit / Withdraw updates

    NestedChain->>NestedNext: Process Sell Orders
    NestedNext->>NestedLiquidity: Execute trades, update virtualPoolSize
    NestedNext->>NestedNext: Record BuyReceipts

    NestedChain->>NestedNext: Process Inbound Liquidity
    NestedNext->>NestedLiquidity: Update virtualPoolSize
    NestedNext->>NestedNext: Assign points

    NestedChain->>NestedNext: Rotate Batches
    NestedNext->>NestedLocked: Become LockedBatch

    %% Nested → Root certificate
    NestedChain->>RootChain: Send certificateResultTx (Nested LockedBatch)

    %% Root reads Nested LockedBatch
    RootChain->>NestedLocked: Read certificateResultTx to unlock trades
    NestedLocked-->>RootChain: Receipt data

    %% Root Batch Processing
    RootChain->>RootLocked: Process LockedBatch
    RootLocked->>RootLiquidity: Success → Move funds
    RootLocked->>UserRoot: Fail → Refund

    RootChain->>RootLocked: Process Local Liquidity
    RootLocked->>RootLiquidity: Deposit / Withdraw updates

    RootChain->>RootNext: Process Sell Orders
    RootNext->>RootLiquidity: Execute trades, update virtualPoolSize
    RootNext->>RootNext: Record BuyReceipts

    RootChain->>RootNext: Process Inbound Liquidity
    RootNext->>RootLiquidity: Update virtualPoolSize
    RootNext->>RootNext: Assign points

    RootChain->>RootNext: Rotate Batches
    RootNext->>RootLocked: Become LockedBatch
```
### Batch Frequency
- A batch is processed **once per block**.
- Nested chain triggers processing during `begin_block`.
- Root chain triggers processing upon receiving a `certificateResultTx`.

---

### Incomplete or Failed Batches
- If a batch cannot be processed due to incomplete cross-chain receipts, processing is aborted and retried in the next trigger.
- A batch may retry up to **3 times**, with each retry corresponding to **5 blocks**.
- After holding a locked batch for 15 blocks without success, the batch is dropped and all orders are refunded.

---

### Order Processing
- Orders within a batch are sorted pseudorandomly using the **hash of the previous block** on the processing chain.
- All computations, including LP points and swaps, use **integer math with floor rounding**.
- Limit orders that do not meet the requested output (`dY < RequestedAmount`) are **rejected**, and funds are **refunded back to the user**.
- Orders are rejected entirely if the relevant pool (local or virtual) has zero liquidity.

---

### One-Sided Deposits & LP Points
- If total LP points `L == 0`, initial deposits are assigned `L = √(x*y)`.
- LP points for deposits and withdrawals are **mirrored across both chains** for symmetry.
- LP points are **burned only after cross-chain confirmation** of withdrawals.

---

### Cross-Chain Token Security
- Token movements are **secured using a HoldingPool**.
- Tokens are moved from `HoldingPool` to `LiquidityPool` **only after cross-chain receipts confirm successful execution**, preventing double-spending or inconsistent states.

---

### Fees
- A standard **0.3% fee** is applied only to successful trades.
- Fees are **added to the liquidity pool** and accrue to LPs.
- Basic transaction fees still apply, even for failed trades.

---

### Pool Edge Cases
- Batches or orders with zero pool reserves (`x == 0` or `y == 0`) are **rejected**.
- Withdrawals also cannot proceed if reserves are zero.

---

### Master/Slave Chain Semantics
- **Nested chain is the master**, Root chain is the slave.
- Root only reads a Nested `certificateResultTx` after it has been submitted.
- Nested will **not process a new batch** until the previous batch has been fully processed and receipts are confirmed.

---

### Timeouts & Refunds
- Each batch retry corresponds to **5 blocks**.
- Orders in a batch that ultimately fails after retries are **refunded to users**, with LPs absorbing any impact.

---

## Cross-Chain Semantics

- The chain maintains a virtual pool that is initialized by counterparty’s last advertised `PoolSize` and is maintained through each operation.
- Inbound receipts finalize success/failure
- Liquidity operations computed on updated virtual pool after trades

---

## Edge Cases & Determinism

- Reject orders if `x == 0` or `y == 0`
- Always integer math with floor rounding
- Orders pseudorandomly sorted by block hash
- Withdrawals processed before deposits
- Deterministic across both chains

---

## Parameters (Defaults)

- `f_taker = 0.003` (30 bps)
- Block timeout: 5 blocks, max 3 retries
- Max batch size: 10,000 orders
- Min order amount: chain-specific dust threshold

---

## Cross-Chain Flow Summary

- **Nested → Root**
    - Nested packages `LockedBatch` → `certificateResultTx` → Root
- **Root Chain**
    - Processes inbound batch
    - No result sent back
- **Trigger Table**

| Step           | Nested Chain                                 | Root Chain                                |
|----------------|---------------------------------------------|-------------------------------------------|
| Trigger        | `begin_block` after own CertificateResult   | `deliver_tx` on incoming certificateResultTx |
| Reads From     | Local + Root state                           | Local + incoming transaction              |
| Sends Result   | Yes (to Root)                                | No                                        |

### PR DESCRIPTION
**Pull request by andrewnguyen22**
closes #241 

  Summary

  This PR implements a Decentralized Exchange (DEX) feature with an event indexing system for the Canopy blockchain. The implementation adds comprehensive DEX functionality including limit orders, liquidity pools, and automated market-making capabilities.

Key Changes:

DEX Implementation

  - New transaction types:
    - TxDexLimitOrder - Place limit orders on the DEX
    - TxDexLiquidityDeposit - Deposit liquidity into pools
    - TxDexLiquidityWithdraw - Withdraw liquidity from pools
  - Core DEX logic (fsm/dex.go:1-630, lib/dex.go:1-305):
    - Order matching engine with 3-block execution cycles
    - Liquidity pool management with pool points
    - Automated market maker (AMM) functionality
    - Pool state tracking and validation
  - Comprehensive testing (fsm/dex_test.go:1-3048, lib/dex_test.go:1-201):
    - 3,048 lines of DEX test coverage
    - Tests for order matching, liquidity operations, edge cases
    - Fuzz testing for validation
  - Documentation (fsm/dex.md:1-385):
    - Complete DEX design and implementation details

  Event System

  - Event indexing (fsm/event.go:1-114, lib/event.go:1-366):
    - Track all transaction events (DEX orders, liquidity changes, swaps, governance, etc.)
    - Event storage and querying capabilities
    - Structured event types for filtering
  - Event tests (fsm/event_test.go:1-466, lib/event_test.go:1-142):
    - 466 lines of comprehensive event testing
    - Validation of event emission and indexing
  - Protobuf definitions (lib/.proto/event.proto:1-103, lib/.proto/dex.proto:1-115):
    - Structured event and DEX message types
    - Generated pb.go files (lib/event.pb.go:1-944, lib/dex.pb.go:1-508)
  - Indexer updates (store/indexer.go):
    - Event indexing by height, type, and address
    - Seek functionality for efficient queries

  Core Infrastructure Changes

  - Transaction handling (fsm/transaction.go:1-56, fsm/message.go:1-132):
    - New message types for DEX operations
    - Enhanced transaction validation with event emission
    - Message helper functions (fsm/message_helpers.go)
  - Protobuf updates (lib/.proto/message.proto, fsm/message.pb.go):
    - New DEX message definitions
    - Updated message structures
  - State management (fsm/state.go:1-120, fsm/account.go:1-105):
    - Pool state tracking and management
    - Account balance updates for DEX operations
    - Committee assignment handling
  - Account updates (fsm/account.pb.go, lib/.proto/account.proto):
    - Pool points tracking for liquidity providers
    - Enhanced account state
  - Governance (fsm/gov.go, fsm/gov_params.go:1-100):
    - Parameter change handling
    - Event emission for governance actions
  - Validator operations (fsm/validator.go:1-12):
    - Event emission for validator state changes
  - Key management (fsm/key.go:1-24):
    - Updated key prefixes for pools and events
  - Error handling (fsm/error.go:1-12, lib/error.go:1-60):
    - New DEX-specific error codes
    - Enhanced error messages
  - Library utilities (lib/util.go:1-39, lib/tx.go:1-55, lib/store.go:1-31):
    - Helper functions for DEX calculations
    - Event-related utilities
    - Storage interface updates

  CLI & RPC Enhancements

  - New admin commands (cmd/cli/admin.go:59-275):
    - tx-dex-limit-order
    - tx-dex-liquidity-deposit
    - tx-dex-liquidity-withdraw
  - Data population tool (cmd/cli/populate.go:1-989):
    - Automated chain population for testing/development
    - Exercises all 13+ transaction types systematically
    - Creates realistic test data for DEX, governance, staking, etc.
  - CLI initialization (cmd/cli/cli.go:43-115):
    - --populate flag for automatic data generation
    - FSM initialization updates
  - New query endpoints (cmd/rpc/query.go:1-99):
    - Query events by height, type, and address
    - DEX pool and order queries
  - Route updates (cmd/rpc/routes.go):
    - Added event query routes
    - Enhanced RPC documentation (README.md expanded by ~600 lines)
  - Admin/client endpoints (cmd/rpc/admin.go:1-57, cmd/rpc/client.go:1-120):
    - DEX transaction submission endpoints
    - Event query handlers

  UI Updates

  - Explorer enhancements (cmd/rpc/web/explorer/):
    - Event display in modal (components/modal.jsx)
    - Batch operation visualization (new batch.png icon)
    - Enhanced table component for event data (components/table.jsx)
    - API integration for events (components/api.js)
  - Wallet improvements (cmd/rpc/web/wallet/):
    - DEX operation support in UI (components/account.jsx)
    - New SVG icons for DEX operations (components/svg_icons.jsx)
    - Enhanced API integration (components/api.js:1-114)
    - Utility functions for event formatting (components/util.js)

  Test Plan

  The populate tool (cmd/cli/populate.go) provides comprehensive test coverage by executing all transaction types including:
  - DEX limit orders, liquidity deposits/withdrawals
  - Event indexing verification
  - Integration with existing swap, governance, and staking operations

### CODE REVIEW

File: .docker/volumes/node_2/genesis.json [hunk 1]
```diff
@@ -14,27 +14,25 @@
       "amount": 1000000
     }
   ],
+  "pools": [
```
Comment (rem1niscence) [inline .docker/volumes/node_2/genesis.json]: isn't the pool ID auto generated based on the chain ID, and is that amount calculated as part of the total supply?

File: cmd/cli/populate.go [hunk 2]
```diff
@@ -0,0 +1,989 @@
...
+}
+
+// =====================================================
+// Query Request Types
+// =====================================================
+type heightRequest struct {
+	Height uint64 `json:"height"`
+}
+
+type chainRequest struct {
```
Comment (rem1niscence) [inline cmd/cli/populate.go]: I'll only use this struct as an example but there's like ~10 `<type>Request` structs unused on this populator, we might want to clean them up if they wont be ultimately used
